### PR TITLE
Introduce source and receiver ID cols

### PR DESCRIPTION
**Pull request by alexanderkuvaev**
- Two new survey attributes are introduced: `source_id_cols` and `receiver_id_cols`, which store names of trace headers that uniquely identify a seismic source or receiver respectively. They can be passed either directly to the `Survey.__init__` or subsequently using `set_source_id_cols` and `set_receiver_id_cols` methods.
- `n_sources` and `n_receivers` properties are also introduced. They are also printed in `info` (wow!)
- All survey plotters now account for these attributes (including `construct_fold_map`, `construct_header_map` and `plot_geometry`),
- Prettify validation of survey headers:
  - Some extra checks were added,
  - Now all checks return the number and fraction of "bad" traces/sources/receivers

### CODE REVIEW

**Review by anton-br — APPROVED** [review]

File: seismicpro/survey/survey.py [hunk 1]
```diff
@@ -1214,84 +1350,132 @@ def plot_geometry(self, **kwargs):
...
         ----------
-        attribute : str
-            If "fold", calculates the number of traces in gathers defined by `by`. Otherwise defines a survey header
-            name to construct a map for.
+        col : str
+            Headers column to extract values from.
         by : {"source", "shot", "receiver", "rec", "cdp", "cmp", "midpoint", "bin", "supergather"}
             Gather type to aggregate header values over.
+        id_cols : str or list of str, optional
+            Trace headers that uniquely identify a gather of the chosen type. Acts as an index of the resulting map.
```
Comment (nikita-klsh) [inline seismicpro/survey/survey.py:1433]: Please mention that by default `id_cols` is being infered now, as this logic affects maps

File: seismicpro/survey/survey.py [hunk 2]
```diff
@@ -633,8 +761,10 @@ def dist_to_bin_contours(self, bins):
         Raises
         ------
         ValueError
-            If survey geometry was not inferred.
+            If properties of survey binning were not inferred.
         """
+        if self.bin_contours is None:
```
Comment (nikita-klsh) [inline seismicpro/survey/survey.py:766]:
```suggestion
        if not self.has_inferred_binning:
```

File: seismicpro/survey/survey.py [hunk 3]
```diff
@@ -1142,26 +1272,32 @@ def generate_supergathers(self, centers=None, origin=None, size=3, step=20, bord
         ------
         KeyError
             If `INLINE_3D` and `CROSSLINE_3D` headers were not loaded.
+        ValueError
+            If supergathers have already been generated.
         """
+        line_cols = ["INLINE_3D", "CROSSLINE_3D"]
+        super_line_cols = ["SUPERGATHER_INLINE_3D", "SUPERGATHER_CROSSLINE_3D"]
+        if not self.has_inferred_binning:
+            raise KeyError("INLINE_3D and CROSSLINE_3D headers must be loaded")
```
Comment (nikita-klsh) [inline seismicpro/survey/survey.py:1281]: You actually check whether the binning was performed

**Review by nikita-klsh — APPROVED** [review]

File: seismicpro/survey/survey.py [hunk 4]
```diff
@@ -1214,84 +1350,132 @@ def plot_geometry(self, **kwargs):
...
+        if drop_duplicates:
+            metric_data.drop_duplicates(inplace=True)
+        index = metric_data[index_cols] if index_cols is not None else None
+        coords = metric_data[coords_cols]
+        values = metric_data[name]
+
+        metric = SurveyAttribute(name=name).provide_context(survey=self)
+        return metric.construct_map(coords, values, index=index, agg=agg, bin_size=bin_size)
+
+    def construct_header_map(self, col, by, id_cols=None, drop_duplicates=False, agg=None, bin_size=None):
```
[the remark this answers is not in the document]
Comment (alexanderkuvaev) [reply seismicpro/survey/survey.py:1413]: Discussed to keep as is for now, but perform a global renaming in future pull requests: almost all objects will be affected if done properly.

File: seismicpro/survey/survey.py (hunk 1 above)
Comment (alexanderkuvaev) [reply seismicpro/survey/survey.py:1433]: Done both here and in `construct_fold_map`

File: seismicpro/survey/survey.py (hunk 2 above)
Comment (alexanderkuvaev) [reply seismicpro/survey/survey.py:766]: Fixed both here and in `dist_to_geographic_contours`

File: seismicpro/survey/survey.py (hunk 3 above)
Comment (alexanderkuvaev) [reply seismicpro/survey/survey.py]: Kept the raise, but updated the check

**Pull request merged**